hbase-client/src/main/java/org/apache/hadoop/hbase/client/SlowLogPayload.java

```diff
@@ -0,0 +1,318 @@
+      this.callDetails = callDetails;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setParam(String param) {
+      this.param = param;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setRegionName(String regionName) {
+      this.regionName = regionName;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setUserName(String userName) {
+      this.userName = userName;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setMultiGetsCount(int multiGetsCount) {
+      this.multiGetsCount = multiGetsCount;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setMultiMutationsCount(int multiMutationsCount) {
+      this.multiMutationsCount = multiMutationsCount;
+      return this;
+    }
+
+    public SlowLogPayloadBuilder setMultiServiceCalls(int multiServiceCalls) {
+      this.multiServiceCalls = multiServiceCalls;
+      return this;
+    }
+
+    public SlowLogPayload build() {
+      return new SlowLogPayload(startTime, processingTime, queueTime, responseSize,
+        clientAddress, serverClass, methodName, callDetails, param, regionName,
+        userName, multiGetsCount, multiMutationsCount, multiServiceCalls);
+    }
+  }
+
+  @Override
+  public boolean equals(Object o) {
+    if (this == o) {
+      return true;
+    }
+
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+
+    SlowLogPayload that = (SlowLogPayload) o;
+
+    if (startTime != that.startTime) {
+      return false;
+    }
+    if (processingTime != that.processingTime) {
+      return false;
+    }
+    if (queueTime != that.queueTime) {
+      return false;
+    }
+    if (responseSize != that.responseSize) {
+      return false;
+    }
+    if (multiGetsCount != that.multiGetsCount) {
+      return false;
+    }
+    if (multiMutationsCount != that.multiMutationsCount) {
+      return false;
+    }
+    if (multiServiceCalls != that.multiServiceCalls) {
+      return false;
+    }
+    if (clientAddress != null ? !clientAddress.equals(that.clientAddress)
+        : that.clientAddress != null) {
+      return false;
+    }
+    if (serverClass != null ? !serverClass.equals(that.serverClass)
+        : that.serverClass != null) {
+      return false;
+    }
+    if (methodName != null ? !methodName.equals(that.methodName) : that.methodName != null) {
+      return false;
+    }
+    if (callDetails != null ? !callDetails.equals(that.callDetails) : that.callDetails != null) {
+      return false;
+    }
+    if (param != null ? !param.equals(that.param) : that.param != null) {
+      return false;
+    }
+    if (userName != null ? !userName.equals(that.userName) : that.userName != null) {
+      return false;
+    }
+    return regionName != null ? regionName.equals(that.regionName) : that.regionName == null;
+  }
+
+  @Override
+  public int hashCode() {
+    int result = (int) (startTime ^ (startTime >>> 32));
```

Comment:
Is this necessary? Can you educate me how you come up with this one? thanks

Reply:
Oh okk, here hashCode() and equals() were just generated by IDE, may be this is the default way of dealing with uniqueness of long values, even `responseSize` has similar formula.

Btw, since I was not sure about this so tried searching for a while and found that this is the exact implementation of `Long.hashCode(x)` method:

```
    /**
     * Returns a hash code for a {@code long} value; compatible with
     * {@code Long.hashCode()}.
     *
     * @param value the value to hash
     * @return a hash code value for a {@code long} value.
     * @since 1.8
     */
    public static int hashCode(long value) {
        return (int)(value ^ (value >>> 32));
    }
```